    

# Practica de refuerzo
## Pipeline de Limpieza de Datos


**Estudiante:** Gerard Almanzar
**Curso:** AI & Machine Learning
**Dataset:** gimnacio_raw.csv

### Objetivo
practica de refurzo para seguir afinando lo que ya aprendido en clases con el dataset del Titanic.





## Sobre el dataset


Es un registro de clientes de un gimnasio con tres sedes en Santo Domingo (Naco, Piantini y Gazcue).

| Columna | Significado |
|---|---|
| ClienteID | Número de identificación único de cada cliente |
| Nombre | Nombre completo del cliente |
| Edad | Edad en años |
| Plan | Tipo de membresía: 1 = Básico, 2 = Premium, 3 = VIP |
| Sucursal | Sede donde se inscribió: Naco, Piantini o Gazcue |
| DiasAsistidosMes | Días que asistió al gimnasio en el último mes |
| Email | Correo electrónico (el gimnasio empezó a pedirlo hace poco) |

### Objetivo
Como cualquier dataset real, llega con información incompleta, algunos registros repetidos por error, y una columna numérica que en realidad es una categoría. Limpiar el dataset y guardar la informacion en un dataset nuevo.



# Módulo 1. Importación de Librerías

En esta sección se importan las librerías necesarias para el análisis y limpieza de datos.

In [2]:
import pandas as pd

#Mantiene 'object' en vez de 'str'
pd.options.future.infer_string = False


# Modulo 2. Carga y Exploración Inicial

Se carga el dataset y se realiza una inspección inicial.

In [8]:
df = pd.read_csv(
    r'C:\Users\user\Desktop\Data\mi-practica-talendig-mes4-001\data\clientes_gimnacio_raw.csv'
)

In [9]:
print("Forma del dataset: ", df.shape)
print("\nPrimeras filas: ")
print(df.head(20))


Forma del dataset:  (63, 7)

Primeras filas: 
    ClienteID               Nombre  Edad  Plan  Sucursal  DiasAsistidosMes  \
0          56     Carlos Fernandez  64.0     1      Naco                 4   
1          17           Juan Reyes  47.0     3      Naco                29   
2           8     Yolanda Castillo  52.0     3      Naco                30   
3          25     Francisco Garcia  43.0     1      Naco                25   
4           7   Patricia Rodriguez   NaN     2      Naco                23   
5          32         Ricardo Pena  43.0     1      Naco                24   
6          37        Yolanda Ortiz  42.0     2      Naco                15   
7          15         Manuel Mejia  51.0     3      Naco                25   
8          46       Esperanza Diaz  52.0     1      Naco                11   
9          11       Sofia Castillo  60.0     2      Naco                25   
10         40          Pedro Perez  96.0     1  Piantini                30   
11         48     

In [10]:
print("\nTipos de datos ")
print(df.info())


Tipos de datos 
<class 'pandas.DataFrame'>
RangeIndex: 63 entries, 0 to 62
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ClienteID         63 non-null     int64  
 1   Nombre            63 non-null     object 
 2   Edad              52 non-null     float64
 3   Plan              63 non-null     int64  
 4   Sucursal          61 non-null     object 
 5   DiasAsistidosMes  63 non-null     int64  
 6   Email             16 non-null     object 
dtypes: float64(1), int64(3), object(3)
memory usage: 3.6+ KB
None


## Primeros Hallazgos



Durante la fase inicial explotaroria de datos se encontraron datos nulos a simple vista tras observar por unos minutos el dataset. Se  encontraron datos incompletos en las Las columnas 'Edad', 'Sucursal' y 'Email'. Tambien se observaron algunos tipos de datos incorrectos dentro del dataset. En el siguiente modulo encontraremos todos los valores nulos de cada columna e iremos trabajandolos con el metodo correspondiente.



# Modulo 3. Diagnostico y Manejo de datos faltantes.

Se identifican y corrigen los valores nulos utilizando media, mediana, moda y valores constantes.

In [11]:
print("Diagnostico de datos faltantes")
print(df.isnull().sum())
print("\nPorcentaje de valores faltantes")
print((df.isnull().sum() / len(df) *100).round(2))


Diagnostico de datos faltantes
ClienteID            0
Nombre               0
Edad                11
Plan                 0
Sucursal             2
DiasAsistidosMes     0
Email               47
dtype: int64

Porcentaje de valores faltantes
ClienteID            0.00
Nombre               0.00
Edad                17.46
Plan                 0.00
Sucursal             3.17
DiasAsistidosMes     0.00
Email               74.60
dtype: float64


## Trabajando con los valores faltantes


In [12]:
df['Edad'] = df['Edad'].fillna(df['Edad'].median())
df['Sucursal'] = df['Sucursal'].fillna(df['Sucursal'].mode()[0])
df = df.drop(columns='Email')
print("\nNulos Restantes:")
print(df.isnull().sum(), df.shape)



Nulos Restantes:
ClienteID           0
Nombre              0
Edad                0
Plan                0
Sucursal            0
DiasAsistidosMes    0
dtype: int64 (63, 6)


# Modulo 4. Detección de Duplicados

In [13]:
#Copiar el dataset a una variable para trabajar con los datos duplicados.

datos_duplicados = df[df.duplicated(keep=False)]

#Ordenar sucursal para ver el nombre, edad y plan.
print("\nFilas antes: ", len(df))
print("\n---Historial completo de duplicados encontrados ordenados por sucursal---")
print(f'\n{datos_duplicados.sort_values(by='Sucursal')}')



Filas antes:  63

---Historial completo de duplicados encontrados ordenados por sucursal---

    ClienteID          Nombre  Edad  Plan Sucursal  DiasAsistidosMes
38         21  Gabriela Mejia  17.0     1   Gazcue                11
50         21  Gabriela Mejia  17.0     1   Gazcue                11
5          32    Ricardo Pena  43.0     1     Naco                24
40         32    Ricardo Pena  43.0     1     Naco                24
52         60   Fernanda Cruz  17.0     3     Naco                13
57         60   Fernanda Cruz  17.0     3     Naco                13


In [14]:
#Procedemos a eliminar los datos duplicados.
df.drop_duplicates(inplace=True)
print(df.duplicated())
print("Filas despues de eliminar duplicados: ", len(df))

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
32    False
33    False
34    False
35    False
36    False
37    False
38    False
39    False
41    False
42    False
43    False
44    False
45    False
46    False
47    False
48    False
49    False
51    False
52    False
53    False
54    False
55    False
56    False
58    False
59    False
60    False
61    False
62    False
dtype: bool
Filas despues de eliminar duplicados:  60


# Modulo 5. Normalización y manejo de tipo de datos.

In [15]:
#Trabajando con tipos de datos
df['Edad'] = df['Edad'].astype(int)
df['Plan'] = df['Plan'].astype('category')
print(df.dtypes)

ClienteID              int64
Nombre                object
Edad                   int64
Plan                category
Sucursal              object
DiasAsistidosMes       int64
dtype: object


In [16]:
df

,ClienteID,Nombre,Edad,Plan,Sucursal,DiasAsistidosMes
0,56,Carlos Fernandez,64,1,Naco,4
1,17,Juan Reyes,47,3,Naco,29
2,8,Yolanda Castillo,52,3,Naco,30
3,25,Francisco Garcia,43,1,Naco,25
4,7,Patricia Rodriguez,47,2,Naco,23
5,32,Ricardo Pena,43,1,Naco,24
6,37,Yolanda Ortiz,42,2,Naco,15
7,15,Manuel Mejia,51,3,Naco,25
8,46,Esperanza Diaz,52,1,Naco,11
9,11,Sofia Castillo,60,2,Naco,25


# Tres preguntas para cerrar

### 1. Mediana vs. promedio, con datos reales
**Calcula la media y la mediana de Edad incluyendo a los clientes mayores de 85 años, y luego calcúlalas otra vez
excluyéndolos. ¿Cuál de las dos medidas cambió más al quitar esos casos? ¿Por qué tiene sentido ese resultado?**

La media cambio mas que la mediana al eliminar los clientes mayores de 85 años. Esto ocurre porque la media utiliza todos los valores del dataframe y es sensible a los valores extremos. En cambio, la mediana representa el valor central de los datos ordenados y apenas se ve afectada por la eliminacion de unos pocos valores altos.

### 2. Dos columnas, dos estrategias distintas
**Sucursal tiene apenas 3.17% de nulos y Email tiene 74.60%. ¿Por qué la estrategia correcta no es la misma para
ambas, aunque las dos sean columnas de texto?**

En la columan 'Sucursal' solo faltan unos pocos datos, asi que podemos completarlo facilmente y seguir aprovechando esa informacion. Sin embargo, en la columna 'Email' faltan la mayoria de los registros. Si intentaramos rellenarlos, estariamos adivinando demasiados datos y eso haria que el analisis fuera poco confiable. Por esta razon, lo mejor es eiliminar esa columna y trabajar con informacion de mejor calidad.

### 3. Mirando hacia adelante
***Si este gimnasio quisiera más adelante predecir qué clientes están en riesgo de cancelar su membresía, ¿qué
precaución tomarías con los datos que acabas de limpiar antes de usarlos para eso?***

Aunque los datos ya estan limpios, antes de usarlos para hacer predicciones verificaria que contengan la informacion necesaria para entrenar el modelo. En este caso seria indispensable contar con mas atributos o features que indiquen, por ejemplo, si el cliente cancelo o no la membresia ya que esa sera la varaiable que el modelo aprendera a predecir. Tambien revisaria que los datos sigan siendo representativos de la realidad y que variables importantes como los dias de asistencia no se hayan perdido durante el preoceso de limpieza.


In [17]:
mediana_original = df['Edad'].median()
media_original = df['Edad'].mean()

df_sin85 = df[df['Edad'] <= 85]

media_sin85 = df_sin85['Edad'].mean()
mediana_sin85 = df_sin85['Edad'].median()

print("----Resultados Con todas las edades------")
print("Media: ", media_original)
print("Mediana: ", mediana_original)

print("\n----Resultados sin edades mayores de 85----")
print("Media: ", media_sin85)
print("Mediana: ", mediana_sin85)

----Resultados Con todas las edades------
Media:  46.5
Mediana:  47.0

----Resultados sin edades mayores de 85----
Media:  44.05263157894737
Mediana:  47.0
